### DISCUSSION TITLE
Customizing Typo Tolerance Settings

### DISCUSSION THREAD

**Discussion by edsonfeimberg** [meilisearch/product#312] [Feedback & Feature Proposal]
I am having problems with with the following rules:

- If the query word is between 1 and 4 characters long, therefore, no typo is allowed. Only documents that contain words that start with or are of equal length with this query word are considered valid for this request.
- If the query word is between 5 and 8 characters long, one typo is allowed. Documents that contain words that match with one typo are retained for the next steps.

In my project the user searches for cars, or car brands. They usually have very short names, and this conditions makes Meilisearch not very effective at guessing what the user wants. Examples:

Wants "jetta" but types jeta.
Wants "volvo" but types vouvu
Wants "gol" but types "gou"

and a final example of a "hard" one:
wants "volkswagen" but types something as "Vowksvagem"

is there something i can do to show better results for this kind of users (poor educational attainment  acessibility)?

**Comment by gmourier** [comment]
Hi @edsonfeimberg 👋, thanks for this feedback and your very pertinent question.

We have already had several requests to customize the behavior of typo tolerance. 

Thinking of it, here is what might be allowed as a first step: 

- It should be possible to disable the typo tolerance
- Define the minimum size of a word for 1 typo
- Define the minimum size of a word for 2 typo

We could also go further and disable the typo tolerance on some fields rather than the whole.

Do you have any other ideas that might be of interest regarding the customization of the typo tolerance behavior?

Unfortunately, I don't have a solution in mind to get around this problem in your specific case.

Note: 
Related to https://github.com/meilisearch/product/discussions/300 and https://github.com/meilisearch/product/discussions/176

**Reply by edsonfeimberg** [reply]
Thank you for you response! The implementation of these 3 items you listed would be an amazing improvement!  Considering the team already had several requests related to this topic, could you share something about the current state of development of those features? Are they yet to be included to the roadmap, already part of one, or if it is already at development stage?

**Reply by gmourier** [reply]
We haven't started at all, we're just collecting feedback at this point! In 2022, we'll focus a bit more on search features. I'll be able to give you more visibility soon as we're in the process of establishing the roadmap for 2022 on the core side. 🔌 ⚡️

**Comment by eikooc** [comment]
Hi. I'd also like to chime in. For my use case I can live with only defining the minimum size of a word for 1 - but customizing it even more or having fuzzy search would also be really nice! :)

**Comment by gmourier** [comment]
Hey folks 👋

Update regarding the customization of the tolerance typo.

For **v0.27.0**, we are aiming to add the ability to customize the tolerance typo within a new setting.

The properties for this setting would be:

- `enabled`: Whether the typo tolerance feature is enabled. (Default: `true`)
- `disableOnAttributes`: Disable the typo tolerance feature on the specified document attributes. (Default: `[]`)
- `disableOnWords`: Disable the typo tolerance feature for a set of query terms given during a search query. (Default: `[]`)
- `minWordSizeFor1Typo`: Customize the minimum size for a word to tolerate 1 typo. (Default: `5`)
- `minWordSizeFor2Typos`: Customize the minimum size for a word to tolerate 2 typos (Default: `9`)

A specification is being written here: https://github.com/meilisearch/specifications/pull/117

We would like to check that the namings chosen for the properties are clear to you. You can directly comment in the specification PR. 

We have thought about adding the typo tolerance object as a search parameter so that it can be customized at search time but this is a subject we will deal with later.  🎯

**Comment by gmourier** [comment]
Hey 👋

The customization of the typo tolerance setting is now available (v0.27 release)✨

📚 [Learn more in our dedicated documentation](https://docs.meilisearch.com/learn/configuration/typo_tolerance.html)

Since I'm locking this one, feel free to share what can be improved by [creating a new discussion](https://github.com/meilisearch/product/discussions/new).

Thanks for being part of the Meilisearch community.
